pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,179 @@
+            raise InvalidDataException(str(error))
+
+        entries = []
+        for repo_id, details in data["repos"].items():
+            entries.append(MaintenanceEntry(id=repo_id, **details))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            entries=tuple(entries),
+        )
+
+        return maintenance
+
+    def _json(self):
+        """Output a json format report.
+
+        Returns:
+            A string contains the maintenance report in json format.
+        """
+        report = {
+            "last_updated": self.last_updated,
+            "last_updated_by": self.last_updated_by,
+            "repos": {},
+        }
+
+        for entry in self.entries:
+            report["repos"].update(
+                {
+                    entry.id: {
+                        "message": entry.message,
+                        "owner": entry.owner,
+                        "started": entry.started,
+                    }
+                }
+            )
+
+        return json.dumps(report, indent=4, sort_keys=True)
+
+    def add(self, repo_ids, **kwargs):
+        """Add entries to maintenance report and update the timestamp. Every
+        entry added to the report represents a repository in maintenace mode.
+
+        Args:
+            repo_ids (list[str]):
+                A list of repository ids. New entries with these repository ids will
+                be added to the maintenance report.
+
+            Optional keyword args:
+                message (str):
+                    Reason why put the repo to maintenance.
+                owner (str):
+                    Who set the maintenance mode.
+
+        """
+        message = kwargs.get("message") or "Maintenance mode is enabled"
+        owner = kwargs.get("owner") or OWNER
+
+        to_add = []
+        for repo in repo_ids:
+            to_add.append(
+                MaintenanceEntry(
+                    id=repo, owner=owner, message=message, started=iso_time_now()
+                )
+            )
+        entries = list(self.entries)
+        entries.extend(to_add)
+
+        return MaintenanceReport(last_updated_by=owner, entries=tuple(entries))
+
+    def remove(self, repo_ids, **kwargs):
+        """Remove entries from the maintenece report. Once the entry is removed,
+        it means the corresponding repository isn't in maintenance mode anymore.
+
+        Args:
+            repo_ids (list[str]):
+                A list of repository ids. Entries match repository ids will be removed
+                from the maintenance report.
+
+            Optional keyword args:
+                owner (str):
+                    Who unset the maintenance mode.
+        """
+        owner = kwargs.get("owner") or OWNER
+
+        repo_ids = set(repo_ids)
+        # convert to set, make checking faster
+        to_remove = []
```

Comment:
I think you could do this in a single loop and (in my opinion) would be cleaner:

```suggestion
        new_entries = []
        for entry in self.entries:
          if entry.id not in repo_ids:
            new_entries.append(entry)

        return attr.evolve(self, entries=tuple(new_entries))
```